collector/pg_extensions.go

```diff
@@ -0,0 +1,112 @@
+package collector
+
+import (
+	"context"
+	"database/sql"
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"strconv"
+)
+
+func init() {
+	registerCollector("extensions", defaultEnabled, NewExtensionsCollector)
+}
+
+var pgExtensions = map[string]*prometheus.Desc{
+	"pg_available_extensions": prometheus.NewDesc(
+		"pg_available_extensions",
+		"Extensions that are available for installation",
+		[]string{
+			"name",
+			"default_version",
+			"installed_version",
+		},
+		prometheus.Labels{},
+	),
+	"pg_extensions": prometheus.NewDesc(
+		"pg_extensions",
+		"Installed extensions",
+		[]string{
+			"extname",
+			"extrelocatable",
+			"extversion",
+		},
+		prometheus.Labels{},
+	),
+}
+
+type ExtensionsCollector struct {
+	logger log.Logger
+}
+
+func NewExtensionsCollector(logger log.Logger) (Collector, error) {
+	return &ExtensionsCollector{logger: logger}, nil
+}
+
+func (e *ExtensionsCollector) Update(ctx context.Context, server *server, ch chan<- prometheus.Metric) error {
+	db, err := server.GetDB()
+	if err != nil {
+		return err
+	}
+
+	err = e.scrapeAvailableExtensions(ctx, db, ch)
+	if err != nil {
+		return err
+	}
+
+	err = e.scrapeInstalledExtensions(ctx, db, ch)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *ExtensionsCollector) scrapeInstalledExtensions(ctx context.Context, db *sql.DB, ch chan<- prometheus.Metric) error {
+	rowsExtensions, err := db.QueryContext(ctx, `SELECT extname, extrelocatable, extversion FROM pg_extension`)
```

Comment:
in what PG versions this table is present? is the structure the same across all versions?

Reply:
pg_extension was added in [PostgreSQL 9.1] (https://pgpedia.info/p/pg_extension.html).

there was added a column oid in 12 version, but we don't use it in the query.

Reply:
cool, thx for checking